### PR TITLE
Bytes optimizations from client (WIP)

### PR DESCRIPTION
**Pull request by nicktindall**
Just a PR to share what I have so far.

### CODE REVIEW

File: src/main/java/net/openhft/chronicle/bytes/AbstractBytes.java [hunk 1]
```diff
@@ -1015,6 +1015,24 @@ protected long prewriteOffsetPositionMoved(long subtracting)
...
+        int currentOffset = off;
+        while (remainingToCopy > 0) {
+            int currentBatchSize = Math.min(remainingToCopy, safeCopySize());
+            long offsetInRDO = readOffsetPositionMoved(currentBatchSize);
+            bytesStore.read(offsetInRDO, bytes, currentOffset, currentBatchSize);
+            currentOffset += currentBatchSize;
+            remainingToCopy -= currentBatchSize;
+        }
+        return totalToCopy;
+    }
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/bytes/AbstractBytes.java:1034]: So this delegates to the BytesStore to do the copy in `safeCopySize()` chunks (not sure why it needs to be chunked, was copying that from write). I guess it gives the BytesStore the opportunity to implement an optimized copy.

File: src/main/java/net/openhft/chronicle/bytes/internal/NativeBytesStore.java [hunk 2]
```diff
@@ -824,16 +825,7 @@ public boolean sharedMemory() {
...
-        final long address = this.address + translate(offsetInRDI);
-        for (i = 0; i < len - 7; i += 8)
-            UnsafeMemory.unsafePutLong(bytes, i, memory.readLong(address + i));
-        if (i < len - 3) {
-            UnsafeMemory.unsafePutInt(bytes, i, memory.readInt(address + i));
-            i += 4;
-        }
-        for (; i < len; i++)
-            UnsafeMemory.unsafePutByte(bytes, i, memory.readByte(address + i));
+        memory.copyMemory(this.address + translate(offsetInRDI), bytes, offset + Unsafe.ARRAY_BYTE_BASE_OFFSET, len);
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/bytes/internal/NativeBytesStore.java]: Now this is where it ends up, it had that bug that offset wasn't respected, this change fixes it, but if you look at the `memory.copyMemory` implementation it ends up doing the same thing that the original implementation did

File: src/main/java/net/openhft/chronicle/bytes/internal/NativeBytesStore.java [hunk 3]
```diff
@@ -26,6 +26,7 @@
 import net.openhft.chronicle.core.util.SimpleCleaner;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
+import sun.misc.Unsafe;
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/bytes/internal/NativeBytesStore.java]: shouldn't import this - should use our unsafe UnsafeMemory

File: src/main/java/net/openhft/chronicle/bytes/internal/NativeBytesStore.java (hunk 2 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/bytes/internal/NativeBytesStore.java]: cool

File: src/main/java/net/openhft/chronicle/bytes/AbstractBytes.java (hunk 1 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/bytes/AbstractBytes.java:1034]: From looking at this some more, I think we may be able to do away with the chunking in this method - but let's review that later today - ah actually, this calls ByteStore.readBytes which does not do chunking. Maybe `net.openhft.chronicle.core.UnsafeMemory#readBytes` should? That is scope creep so let's leave

**Review by JerryShea — APPROVED** [review]

File: src/main/java/net/openhft/chronicle/bytes/AbstractBytes.java [hunk 4]
```diff
@@ -1015,6 +1015,29 @@ protected long prewriteOffsetPositionMoved(long subtracting)
...
+
+    @Override
+    public long read(long offsetInRDI, @NotNull byte[] bytes, int offset, int length) throws IllegalStateException {
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/bytes/AbstractBytes.java]: needed to add this here. I didn't chunk it

**Review by JerryShea — APPROVED** [review]

File: src/main/java/net/openhft/chronicle/bytes/RandomDataInput.java [hunk 5]
```diff
@@ -584,7 +584,7 @@ default boolean compareUtf8(long offset, @Nullable CharSequence other)
         return BytesInternal.toByteArray(this);
     }
 
-    default long read(long offsetInRDI, byte[] bytes, int offset, int length)
+    default long read(long offsetInRDI, @NotNull byte[] bytes, int offset, int length)
```
Comment (minborg) [inline src/main/java/net/openhft/chronicle/bytes/RandomDataInput.java]: `@NotNull` for arrays means that the elements in the array cannot be null. Consequently, this annotation for primitive arrays makes no sense.

File: src/test/java/net/openhft/chronicle/bytes/BytesTest.java [hunk 6]
```diff
@@ -1016,4 +1016,17 @@ public void testAppendReallySmallDouble() {
             assertEquals(d, bytes.parseDouble(), err);
         }
     }
-}
\ No newline at end of file
+
+    @Test
```
Comment (minborg) [inline src/test/java/net/openhft/chronicle/bytes/BytesTest.java:1020]: +1 for testing

File: src/main/java/net/openhft/chronicle/bytes/AbstractBytes.java [hunk 7]
```diff
@@ -1015,6 +1015,29 @@ protected long prewriteOffsetPositionMoved(long subtracting)
         return this;
     }
 
+    @Override
+    public int read(@NotNull byte[] bytes, int off, int len) throws BufferUnderflowException, IllegalStateException {
```
Comment (minborg) [inline src/main/java/net/openhft/chronicle/bytes/AbstractBytes.java:1019]: I suggest `byteArray` over `bytes`. After all, we are in the `Bytes` class.

**Review by minborg — APPROVED** [review]

**Pull request merged**